### PR TITLE
[exporter/file] Add group_by configuration

### PR DESCRIPTION
**Pull request by adam-kiss-sg**
**Description:** 

Added the option to write telemetry data into multiple files, where the file path is based on a resource attribute.

**Link to tracking Issue:** 

#24654

**Testing:** 

Added tests and benchmark for new functionality.

**Documentation:**

Updated README.md

### CODE REVIEW

File: exporter/fileexporter/grouping_file_exporter.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+			continue
+		}
+
+		err = e.write(ctx, pathSegment, buf)
+		if err != nil {
+			errs = errors.Join(errs, err)
+		}
+	}
+
+	return errs
```
Comment (adam-kiss-sg) [inline exporter/fileexporter/grouping_file_exporter.go:68]: The most likely errors here are permission errors when creating directories or writing to files.

What's the result of returning this error from the exporter? Is the error propagated back, or is it retried? Is there any way of marking this error as non-retryable, so the collector simply logs it as an error and moves on?

File: exporter/fileexporter/README.md [hunk 2]
```diff
@@ -78,6 +83,15 @@ When `format` is json and `compression` is none , telemetry data is written to f
 
 Otherwise, when using `proto` format or any kind of encoding, each encoded object is preceded by 4 bytes (an unsigned 32 bit integer) which represent the number of bytes contained in the encoded object.When we need read the messages back in, we read the size, then read the bytes into a separate buffer, then parse from that buffer.
 
+## Group by attribute
+
+By specifying `group_by.resource_attribute` in the config, the exporter will determine a filepath for each telemetry record, by concatenating the value of the resource attribute to the `path` configuration value.
```
Comment (djaglowski) [inline exporter/fileexporter/README.md:88]: Technically not concatenation with the updated design.

```suggestion
By specifying `group_by.resource_attribute` in the config, the exporter will determine a filepath for each telemetry record, by substituting the value of the resource attribute into the `path` configuration value.
```

File: exporter/fileexporter/README.md [hunk 3]
```diff
@@ -78,6 +83,15 @@ When `format` is json and `compression` is none , telemetry data is written to f
...
+
+The exporter will NOT write files outside of the `path` config value, eg: if `path` is "/data", and the resource attribute value is "../etc/my_config.yaml", then the final path will be sanitized to "/data/etc/my_config.yaml".
```
Comment (djaglowski) [inline exporter/fileexporter/README.md:90]: This example may need to be updated based on the newer design.

File: exporter/fileexporter/config.go [hunk 4]
```diff
@@ -63,6 +68,21 @@ type Rotation struct {
...
+	Enabled bool `mapstructure:"enabled"`
+
+	// ResourceAttribute specifies the name of the resource attribute that
+	// contains the path segment of the file to write to. The final path will be
+	// the Path config value, with the * replaced with the value of this resource
+	// attribute. Default is "fileexporter.path_segment".
+	ResourceAttribute string `mapstructure:"resource_attribute"`
+
+	// MaxOpenFiles specifies the maximum number of open file descriptors for the output files.
+	// The fefaults is 100.
```
Comment (djaglowski) [inline exporter/fileexporter/config.go:82]:
```suggestion
	// The default is 100.
```

File: exporter/fileexporter/grouping_file_exporter.go (hunk 1 above)
Comment (djaglowski) [reply exporter/fileexporter/grouping_file_exporter.go:68]: Good question. Maybe @dmitryax knows?

File: exporter/fileexporter/grouping_file_exporter.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+	fullPath := path.Clean(e.pathPrefix + pathSegment + e.pathSuffix)
+	if strings.HasPrefix(fullPath, e.pathPrefix) {
+		return fullPath
+	}
+
+	// avoid path traversal vulnerability
+	return path.Join(e.pathPrefix, path.Join("/", pathSegment+e.pathSuffix))
+}
+
+func (e *groupingFileExporter) onEnvict(_ string, writer *fileWriter) {
```
Comment (djaglowski) [inline exporter/fileexporter/grouping_file_exporter.go:209]:
```suggestion
func (e *groupingFileExporter) onEvict(_ string, writer *fileWriter) {
```

**Review by djaglowski — COMMENTED** [review]
Generally looks good to me

File: exporter/fileexporter/README.md [hunk 6]
```diff
@@ -25,9 +25,9 @@ Exporter supports the following features：
 
 + Support for compressing the telemetry data before exporting.
 
++ Supports for writing into multiple files, where the file path is determined by a resource attribute.
 
 Please note that there is no guarantee that exact field names will remain stable.
-This intended for primarily for debugging Collector without setting up backends.
```
Comment (astencel-sumo) [inline exporter/fileexporter/README.md:30]: Why is this line removed in this PR? Does not seem related to the change.

Comment (adam-kiss-sg) [reply exporter/fileexporter/README.md:30]: With every new feature added to this exporter, it becomes less and less of a tool for debugging, so the removal of this line is lightly related to the new functionality added in this pr. That was my reason for removing it (after discussing it in the related issue).

But I'm not going to argue about a line in a readme. If you feel strongly about keeping it (or not removing it in this PR because it's not strictly related to it), I'm happy to add it back.

Comment (astencel-sumo) [reply exporter/fileexporter/README.md:30]: > after discussing it in the related issue

Thanks, this is a good point. I'm fine with this. :+1:

File: exporter/fileexporter/file_exporter_test.go [hunk 7]
```diff
@@ -126,7 +126,7 @@ func TestFileTracesExporter(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			conf := tt.args.conf
-			feI, err := newFileExporter(conf)
+			feI, err := newFileExporter(conf, nil)
```
Comment (djaglowski) [inline exporter/fileexporter/file_exporter_test.go:129]:
```suggestion
			feI, err := newFileExporter(conf, zap.NewNop())
```

File: exporter/fileexporter/file_exporter_test.go [hunk 8]
```diff
@@ -648,7 +648,7 @@ func TestFlushing(t *testing.T) {
 	// Wrap the buffer with the buffered writer closer that implements flush() method.
 	bwc := newBufferedWriteCloser(buf)
 	// Create a file exporter with flushing enabled.
-	feI, err := newFileExporter(cfg)
+	feI, err := newFileExporter(cfg, nil)
```
Comment (djaglowski) [inline exporter/fileexporter/file_exporter_test.go:651]:
```suggestion
	feI, err := newFileExporter(cfg, zap.NewNop())
```

**Review by djaglowski — COMMENTED** [review]
A couple more nits, and I don't have an answer for the permanent/retryable error question, but otherwise looks good.

File: exporter/fileexporter/grouping_file_exporter.go (hunk 1 above)
Comment (astencel-sumo) [reply exporter/fileexporter/grouping_file_exporter.go:68]: Here's my understanding, correct me if I'm wrong.

The file exporter seems to not be using the backoff/retry or queueing functionality from exporterhelper, so this error will be simply returned to the previous component in the pipeline (a processor or a receiver). The error is not wrapped into a [permanent](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.95.0/consumer/consumererror/permanent.go) error with NewPermanent(errs), so the preceding component might or might not retry sending the data, depending on its configuration and/or implementation.

I suppose that's the right way to do. If you are certain the error is permanent, you might want to wrap it with `NewPermanent`.

Comment (adam-kiss-sg) [reply exporter/fileexporter/grouping_file_exporter.go:68]: Thanks @astencel-sumo 

I did a quick search in the code base for "errors.Join", to find any other instances where a component might have partial errors (eg: some logs are processed properly, others are not).

I didn't find anything indicating that these partial errors are handled (eg: only the failed logs would be retried).

In some exporters (eg: awscloudwatchlogsexporter) I see it uses retry/queue using the erporterhelper that you mentioned. In some cases I think that might result in duplicate logs. The chance of a partial failure might be slim with awscloudwatchlogsexporter, but it is much higher with this fileexporter.

Anyway, considering most other exporters/connectors simply return these joined errors as it is here, I would leave it as it is. And I think adding queue/backoff/retry capabilities is outside the scope of this pr.

So, what do you think about merging this as it is, @astencel-sumo , @djaglowski ?

Comment (dmitryax) [reply exporter/fileexporter/grouping_file_exporter.go:68]: Retriable errors can embed the portion of the data that can be retried to avoid double publishing. But, in this case I believe the error is permanent, so just wrap it in https://github.com/open-telemetry/opentelemetry-collector/blob/fb912b97c0e98b380f9ebf9e70e81ece23e3fdb2/consumer/consumererror/permanent.go#L16C6-L16C18

File: exporter/fileexporter/grouping_file_exporter.go [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+
+		buf, err := e.marshaller.marshalMetrics(metrics)
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		err = e.write(ctx, pathSegment, buf)
+		if err != nil {
+			errs = errors.Join(errs, err)
```
Comment (djaglowski) [inline exporter/fileexporter/grouping_file_exporter.go:104]: Permanent error here as well.

File: exporter/fileexporter/grouping_file_exporter.go [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+
+		buf, err := e.marshaller.marshalLogs(logs)
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		err = e.write(ctx, pathSegment, buf)
+		if err != nil {
+			errs = errors.Join(errs, err)
```
Comment (djaglowski) [inline exporter/fileexporter/grouping_file_exporter.go:138]: Permanent error here as well.

File: exporter/fileexporter/grouping_file_exporter.go (hunk 9 above)
Comment (adam-kiss-sg) [reply exporter/fileexporter/grouping_file_exporter.go:104]: oops, long day

Comment (adam-kiss-sg) [reply exporter/fileexporter/grouping_file_exporter.go:104]: The correct place for wrapping it with NewPermanent is a couple lines below, see:

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31396/files#diff-2c93a6fc48837207b37ca4f2224563b7b131a435089dbdbe5acd165318758c42R109

Comment (djaglowski) [reply exporter/fileexporter/grouping_file_exporter.go:104]: Ah, thanks.

**Review by djaglowski — APPROVED** [review]

**Comment by djaglowski** [comment]
Not sure why the checks aren't running. Mind pushing an empty commit to retrigger them?

**Comment by adam-kiss-sg** [comment]
I did a main merge because there were some conflicts. The two failing checks seems to be due to an upstream issue: https://github.com/tcort/markdown-link-check/issues/297

**Review by djaglowski — APPROVED** [review]
